src/ucp/core/ucp_worker.c

```diff
@@ -2385,6 +2385,75 @@ static void ucp_worker_set_max_am_header(ucp_worker_h worker)
                             ucs_min(max_am_header, UINT32_MAX) : 0ul;
 }
 
+void ucp_worker_track_ep_usage(ucp_request_t *req)
+{
+    ucp_worker_h worker = req->send.ep->worker;
+
+    if ((worker->usage_tracker.iter_count++ %
+         UCP_WORKER_PROGRESS_TIMER_SKIP_COUNT) != 0) {
+        return;
+    }
+
+    if (ucs_likely((ucs_get_time() - worker->usage_tracker.last_round) <
+                   worker->context->config.ext.usage_tracker_interval)) {
+        return;
+    }
+
+    if (req->flags & UCP_REQUEST_FLAG_USAGE_TRACKED) {
+        return;
+    }
+
+    req->flags |= UCP_REQUEST_FLAG_USAGE_TRACKED;
+    ucs_usage_tracker_touch_key(worker->usage_tracker.handle, req->send.ep);
+    worker->usage_tracker.samples_count++;
+
+    if ((worker->usage_tracker.samples_count % 100000) == 0) {
+        ucs_usage_tracker_progress(worker->usage_tracker.handle);
+        worker->usage_tracker.last_round = ucs_get_time();
+    }
+}
+
+static ucs_status_t ucp_worker_create_usage_tracker(ucp_worker_h worker)
+{
+    ucs_usage_tracker_params_t params = {0};
+    ucs_status_t status;
+
+    if (!worker->context->config.ext.usage_tracker_enable) {
+        return UCS_OK;
+    }
+
+    params.promote_capacity = 20;
+    params.promote_thresh   = 10;
+    params.remove_thresh    = 0.2;
+    params.exp_decay.m      = 0.8;
+    params.exp_decay.c      = 0.2;
+    params.promote_cb       =
+            (ucs_usage_tracker_elem_update_cb_t)ucs_empty_function;
+    params.demote_cb        =
+            (ucs_usage_tracker_elem_update_cb_t)ucs_empty_function;
+
+    status = ucs_usage_tracker_create(&params, &worker->usage_tracker.handle);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    worker->usage_tracker.iter_count    = 0;
+    worker->usage_tracker.samples_count = 0;
+    worker->usage_tracker.last_round    = ucs_get_time();
+    return UCS_OK;
+}
+
+static void ucp_worker_destroy_usage_tracker(ucp_worker_h worker)
+{
+    if (!worker->context->config.ext.usage_tracker_enable) {
+        return;
+    }
+
+    ucs_usage_tracker_destroy(worker->usage_tracker.handle);
+    uct_worker_progress_unregister_safe(worker->uct,
```

Comment:
Where is the call to `register`?

Reply:
Unresolved

Reply:
I already removed this function call. why unresolved?